### PR TITLE
Add intel_mp phy down alarm

### PR DESCRIPTION
**Pull request by benagricola**
```
 - Adds all mandatory fields from ietf-alarms to alarm-notifications
 - Fixes a bug with `snabb ptree` alarm-notification handling when socket is closed
 - Implements `string-list` alarm codec support for `alt-resource`
 - Modifies alarm classes to support boolean alarms as well as rate-based alarms
```

@dpino as I found / fixed these issues in the process of implementing the `phy-down` alert for the `intel_mp` driver, I've left those changes (and the ones to the alarm classes) in the PR for context.

Happy to remove these from the PR if you think they should be submitted separately and this one should just be 'add fields to alarm-notifications'.

### CODE REVIEW

File: src/apps/intel_mp/intel_mp.lua [hunk 1]
```diff
@@ -446,9 +447,31 @@ function Intel:new (conf)
...
+            alarm_text='Interface is down',
+         }
+      }
+      self.phy_down_state_alarm = CallbackAlarm.new(phy_down_state,
+         1, function()
+            -- Alarm is not active when link is active
+            return not self:link_status()
+         end)
+   end
+
```
Comment (dpino) [inline src/apps/intel_mp/intel_mp.lua]: I think you should be able to code this behavior without implementing a new `CallbackAlarm` for boolean values. I explain how.

Current `CallbackAlarm` takes 4 parameters: `alarm`, `period`, `limit` and a `callback`. Argument `limit` is the value that determines whether the alarm should be raised or cleared. So if the value of the callback is a boolean you can convert the result to an integer (1 or 0) and set `limit` as 0. Something like:

```lua
self.phy_down_state_alarm = CallbackAlarm.new(phy_down_state,
   1, 0, function()
   -- Alarm is not active when link is active
   if not self:link_status() then return 1 else return 0 end
end)
```
So when the value returned is 1, since it's higher than `limit` the alarm is raised. The alarm is cleared otherwise.

File: src/apps/intel_mp/intel_mp.lua [hunk 2]
```diff
@@ -693,6 +716,10 @@ function Intel:pull ()
    if self.run_stats and self.sync_timer() then
       self:sync_stats()
    end
+
+   if self.master and pkts < 1 then
+      self.phy_down_state_alarm:check()
```
Comment (dpino) [inline src/apps/intel_mp/intel_mp.lua]: Bear in mind your alarm if only going to be checked if you receive no packets. I'm thinking of the following scenario:

You run the NIC but receive no packets, then your alarm is checked and it will be raised if the link is down. But then if the link recovers you start receiving packets, however the alarm is not checked again, so it doesn't have a chance to clear.

If the alarm purpose is to report the link state I think it should be checked whether packets were received or not.

File: src/lib/ptree/ptree.lua [hunk 3]
```diff
@@ -225,7 +225,7 @@ end
 function Manager:handle_notification_peer(peer)
    local q = queue.new()
    self.notification_peers[q] = true
-   function q:close()
+   function q.close()
```
Comment (dpino) [inline src/lib/ptree/ptree.lua:228]: This change is not needed. I explain why.

In Lua when an object is called using ':' the caller object is passed as an implicit parameter (that makes possible to reference `self` inside the callee function). The method definition can either use ':' or '.'. If using ':' there's nothing else to do, all good. If the method definition uses '.', then it generally means it doesn't expect any arguments, but it could. Example:

```lua
local t = {}

function t:foo(arg)
   if self then print('self is implicit') end
   print('foo: '..arg)
end
function t.bar(caller, arg)
   if not self then print("self doesn't exist") end
   if caller then print("caller exists") end
   print('bar: '..arg)
end

print(t:foo(42))
print(t:bar(42))
```
Whether to use ':' or '.' in an object method definition which doesn't have arguments depend very much on each person coding style. I generally prefer to use ':' in case the method is extended with arguments in the future. But as for "class methods", such as `new` I prefer to use '.'.

**Review by dpino — CHANGES_REQUESTED** [review]
I still need to review the alt_resource changes, but at first they look good.

File: src/apps/intel_mp/intel_mp.lua (hunk 1 above)
Comment (benagricola) [reply src/apps/intel_mp/intel_mp.lua]: Yep, fair point - I'll give that a shot

File: src/apps/intel_mp/intel_mp.lua (hunk 2 above)
Comment (benagricola) [reply src/apps/intel_mp/intel_mp.lua]: Also a good point - the idea behind checking if pkts < 1 was to avoid checking the link_status register every breath but obviously that doesn't work to detect the link coming back up again. Will have a think about this...

File: src/lib/ptree/ptree.lua (hunk 3 above)
Comment (benagricola) [reply src/lib/ptree/ptree.lua:228]: Yeah so using `:` means `self` is passed through implicitly from the called object - rather than explicitly. 

The problem with using `:` on `q:close()` is that `self.notification_peers` (where `self` is `Manager`)  is referenced *inside* the `close()` function. At that point, if using `:`, `self` is `queue` rather than `Manager` and does not have a `notification_peers` var.

Since the `close` method on `q` doesn't actually refer to itself in any manner, it should be safe to use `q.close` which leaves `self` referencing `Manager`.

File: src/apps/intel_mp/intel_mp.lua (hunk 1 above)
Comment (benagricola) [reply src/apps/intel_mp/intel_mp.lua]: Hi @dpino I remember why this doesn't work now - the code in `Alarm:check()` calculates a difference (rate) based on `value - last_value` - this means that the alarm will automatically resolve itself the second time it runs (as 1 - 1 = 0 and 0 - 0 = 0) if the state has not changed.

Obviously I could use `self.alarm:raise()` and `:clear()` manually without using the Alarm classes but that seems messy - thoughts?

File: src/lib/ptree/ptree.lua (hunk 3 above)
Comment (dpino) [reply src/lib/ptree/ptree.lua:228]: OK, makes perfect sense to me. Since the diff didn't show the body of the function I wrongly assumed it was aesthetic change. So I think this is the fix for the `attempt to index field 'notification_peers'` you mentioned in Slack.

Comment (benagricola) [reply src/lib/ptree/ptree.lua:228]: Yep, should've mentioned that 😃 Fixes the index issue.

File: src/lib/yang/alarms.lua [hunk 4]
```diff
@@ -348,11 +352,20 @@ end
...
    local ret = assert(alarm_list:retrieve(key, args), 'Not supported alarm')
+
+   local alt_resource = args.alt_resource or ret.alt_resource or {}
+   assert(type(alt_resource) == "table")
+
    local status = {
       time = format_date_as_iso_8601(),
       perceived_severity = args.perceived_severity or ret.perceived_severity,
       alarm_text = args.alarm_text or ret.alarm_text,
+      resource      = key.resource,
```
Comment (dpino) [inline src/lib/yang/alarms.lua:363]: An alarm-status-change is defined in the YANG schema as (https://www.ietf.org/archive/id/draft-vallin-netmod-alarm-module-02.txt):

```
+--ro status-change* [time] {alarm-history}?
       |     |  +--ro time                  yang:date-and-time
       |     |  +--ro perceived-severity    severity-with-clear
       |     |  +--ro alarm-text            alarm-text
```

If you add extra information here, I'm not sure it will be shown in those tools that serialize/deserialize the data according to the schema (for instance, `alarms get`). In any case, this change doesn't respect the schema, so it's not correct.

Notice the `alt-resource` list gets automatically copied to the alarm when the alarm is created.

File: src/lib/yang/alarms.lua [hunk 5]
```diff
@@ -348,11 +352,20 @@ end
 -- status change is added to the alarm.
 local function new_alarm (key, args)
    local ret = assert(alarm_list:retrieve(key, args), 'Not supported alarm')
+
+   local alt_resource = args.alt_resource or ret.alt_resource or {}
```
Comment (dpino) [inline src/lib/yang/alarms.lua:356]: According to the YANG schema `alt-resource` is an element of alarm:

```
container alarm-list {
   list alarm {
      key "resource alarm-type-id alarm-type-qualifier";
      uses common-alarm-parameters;
      ...
   }
}
grouping common-alarm-parameters {
   leaf resource {...}
   leaf alarm-type-id {...}
   leaf alarm-type-qualifier {...}
   leaf-list alt-resource {
      ...
   }
}
```

The method `alarm_list:retrieve(key, args)` returns an existing alarm, and if the alarm doesn't exist it's created. All arguments in `args` are copied to the alarm. So if `declare-alarm` is initialized with an `alt-resource` as argument it will be copied to an alarm.

File: src/lib/yang/alarms.lua [hunk 6]
```diff
@@ -389,10 +402,17 @@ end
...
+      assert(type(alt_resource) == "table")
+
       local status = {
          time = assert(format_date_as_iso_8601()),
          perceived_severity = assert(args.perceived_severity or alarm.perceived_severity),
          alarm_text = assert(args.alarm_text or alarm.alarm_text),
+         resource      = key.resource,
+         alarm_type_id = key.alarm_type_id,
+         alarm_type_qualifier = key.alarm_type_qualifier,
+         alt_resource = alt_resource,
```
Comment (dpino) [inline src/lib/yang/alarms.lua:415]: These changes are not correct because they don't respect the semantics of the alarms YANG schema.

File: src/lib/yang/alarms.lua [hunk 7]
```diff
@@ -729,6 +749,7 @@ function selftest ()
    do_add_to_inventory({alarm_type_id='arp-resolution'}, {
       resource='nic-v4',
       has_clear=true,
+      alt_resource={'nic-v4-2'},
```
Comment (dpino) [inline src/lib/yang/alarms.lua:752]: This is OK but you still need to add a test that verifies that when `arp-resolution` alarm is retrieved it contains an `alt_resource` attribute and that attribute is a list. Something like:

```lua
   -- Raising an alarm when alarms is empty, creates an alarm.
   local key = alarm_keys:fetch('nic-v4', 'arp-resolution')
   raise_alarm(key)
   local alarm = assert(state.alarm_list.alarm[key])
   assert(#alarm.alt_resource == 1)
```

File: src/lib/yang/alarms.lua [hunk 8]
```diff
@@ -43,6 +43,10 @@ function notifications ()
    local ret = {}
    local notifications = state.notifications
    for k,v in pairs(notifications.alarm) do
+      -- TODO: Improve JSON encode / decode to support nested tables
+      if type(v.alt_resource) == 'table' then
+         v.alt_resource = table.concat(v.alt_resource, ',')
+      end
```
Comment (dpino) [inline src/lib/yang/alarms.lua:49]: I'm not sure about this change. In one hand, the encode/decode code for alt-resource serialization seems OK to me. It works with a native Lua table. But here `alt_resource` is encoded to a string of separated by commas, but only for notifications.

An `alarm-notification` is defined in the alarms schema as:

```
notification alarm-notification {
        uses common-alarm-parameters;
        uses alarm-state-change-parameters;
}
```

Basically has the same attributes and structure as an alarm. If things are working ok when ptree deserializes notifications to a json object it should deserialize all attributes (https://github.com/Igalia/snabb/blob/lwaftr/src/lib/ptree/ptree.lua#L234).

File: src/lib/ptree/alarm_codec.lua [hunk 9]
```diff
@@ -288,12 +313,12 @@ function selftest ()
...
       local key = {resource='res1', alarm_type_id='type1', alarm_type_qualifier=''}
-      local args = {perceived_severity='critical'}
+      local args = {perceived_severity='critical', alt_resource={'res2'}}
 
       local resource, alarm_type_id, alarm_type_qualifier = alarm:normalize_key(key)
-      local perceived_severity, alarm_text = alarm:normalize_args(args)
+      local perceived_severity, alarm_text, alt_resource = alarm:normalize_args(args)
       local alarm = {resource, alarm_type_id, alarm_type_qualifier,
-                     perceived_severity, alarm_text}
+                     perceived_severity, alarm_text, alt_resource}
```
Comment (dpino) [inline src/lib/ptree/alarm_codec.lua:321]: All this code seems OK to me. However, after reviewing the changes in `lib/yang/alarms.lua` I'm not sure it's working as expected. After this change, without any change in `lib.yang.alarms`, you should see that your `phy-down` alarm has an `alt-resource` list of alarms when is raised. You can query the list of raised alarms using `sudo ./snabb alarms get-state /'.

I also noticed that the declaration of your `phy-down` alarm doesn't have an `alt-resource` attribute.

**Review by dpino — CHANGES_REQUESTED** [review]
The encode/decode serialization code seems OK to me, but I have doubts in several changes in `lib/yang/alarms.lua` (see comments). Definitely the changes that don't respect the alarms schema semantics are not correct.

File: src/lib/yang/alarms.lua (hunk 4 above)
Comment (benagricola) [reply src/lib/yang/alarms.lua:363]: Okay - I think I understand now - I added these to the status item as, as far as I could tell, this is what is outputted by `alarms listen` - I was under the impression that the output of `alarms listen` was an `alarm-notification` but as you point out, it's currently a `status-change` item.

Which makes me think that there's missing functionality here - when a `status-change` occurs and is added to the list, an `alarm-notification` should be sent which contains the full information required. Does this sound correct? Is `alarms listen` *designed* to only send a `status-change` and if so, what's the correct way to work with it to retrieve the resource(s) affected?

Comment (dpino) [reply src/lib/yang/alarms.lua:363]: `alarm listen` is the right command for listening for notifications. `alarm get-state` is like `config state`, it returns state information of the lwAFTR instance but in this case it only reports alarms.

According to the schema, there are three types of notifications:
   * alarm-notification
   * alarm-inventory-changed
   * operator-action

An alarm-notification is sent in the following scenarios:
   * When an alarm is raised.
   * When an alarm is cleared.
   * When there's a status-change in the alarm (for example, because the alarm changes it severity).

Point is since you need a lwAFTR instance before running `alarms listen`, if your alarms are raised between the time you started the lwAFTR and run the listen command, you will miss those notifications. To test whether a notification is sent or not maybe what you could try is to fix the condition that raised your alarm (so it should get cleared and send a notification). I don't know if it's feasible to do that with the link_down alarm you added.

Another possibility would be to add a print in the `add_alarm_notification` and see if it's being executed (definitely not the best way to test it).

As far as I recall, I tested `alarm-inventory-changed` and `operator-action` and checked they were working, but I can't tell now about `alarm-notification`. Maybe there's an issue going on there.

About the info that is going to print out, it would be all the info that is available in the alarm I think. There are fields that are specified in the schema but I didn't support (such as `alt-resource`, but there are others as `related-alarm`, `impacted-resource` or `root-cause-resource`). IMHO the alarms schema is overly complicated and with such a lot of fields per alarm, encoding/decoding of alarms got more complicated too. That's why I initially decided to support only a limited amount of fields (the most fundamental). with the idea that if someone needs more fields in the future they could be added.

Comment (benagricola) [reply src/lib/yang/alarms.lua:363]: I do get JSON output from `alarms listen` for every event, so that at least is working correctly - what I meant was that the JSON emitted from `alarms listen` is technically the `status-change` rather than the `alarm-notification` event. 

The `status` table is what's passed to `new_notification` (in `add_alarm_notification`), so if all of the mandatory fields for `alarm-notification` do not exist in the `status` table then the JSON output does not match the yang `alarm-notification` schema - but if I add the fields to the `status` table then the `status-change` list doesn't match the schema 😄 

Basically, `status-change` and `alarm-notification` are not the same thing but currently both rely on a single table passed in to `add_status_change`.

I'll have a think about how to work around this.

Comment (dpino) [reply src/lib/yang/alarms.lua:363]: You're right, great investigation. I think the issue can be solved if `add_alarm_notification` is passed an `alarm` and not an `status`. Could you give that a try?

Comment (dpino) [reply src/lib/yang/alarms.lua:363]: And there's an extra thing as well. Perhaps it's necessary to change `new_notification` to do a deep-copy of the `value` passed.

**Comment by benagricola** [comment]
Closing in favour of separate PRs
